### PR TITLE
adapter: Limit number of concurrent connections

### PR DESCRIPTION
**Pull request by andrewrodriguez-m**
Proposed fix for https://github.com/MaterializeInc/materialize/issues/18341. A little simpler than the previous efforts (#18702, #19023)

### Motivation
This, along with the other proposed fixes in #18341, is a proxy for "bad things happening"

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/adapter/src/coord/command_handler.rs [hunk 1]
```diff
@@ -207,6 +208,17 @@ impl Coordinator {
         cancel_tx: Arc<watch::Sender<Canceled>>,
         tx: oneshot::Sender<Response<StartupResponse>>,
     ) {
+        if !session.user().is_superuser()
+            && self.active_conns.len()
+                > usize::cast_from(self.catalog().system_config().max_connections())
+        {
+            let _ = tx.send(Response {
+                result: Err(AdapterError::TooManyConnections),
```
Comment (jkosh44) [inline src/adapter/src/coord/command_handler.rs:216]: Is it possible to re-use the `ResourceExhaustion` error here?

**Review by jkosh44 — APPROVED** [review]
LGTM

File: src/adapter/src/coord/command_handler.rs [hunk 2]
```diff
@@ -207,6 +207,25 @@ impl Coordinator {
...
+            ) {
+                Ok(()) => {}
+                Err(e) => {
+                    let _ = tx.send(Response {
+                        result: Err(e),
+                        session,
+                    });
+                    return;
+                }
+            }
```
Comment (ParkMyCar) [inline src/adapter/src/coord/command_handler.rs:226]: no need to make this change, you could also rewrite this with a `if let`, e.g.
```
if let Err(e) = self.validate_resource_limit(...) {
  let _ = tx.send(...);
  return;
}
```
this does nothing if the result is `Ok(...)` like we're currently doing

Comment (andrewrodriguez-m) [reply src/adapter/src/coord/command_handler.rs]: done

**Pull request merged**